plugins/woocommerce/includes/wc-product-functions.php

```diff
@@ -1834,3 +1835,34 @@ function wc_product_attach_featured_image( $attachment_id, $product = null, $sav
 	}
 }
 add_action( 'add_attachment', 'wc_product_attach_featured_image' );
+
+/**
+ * Delete related product transients when a product is saved.
+ * This is necessary because changing one product might affect many related products.
+ *
+ * @since 9.8.1
+ * @param int $product_id Product ID.
+ */
+function wc_delete_related_product_transients( $product_id ) {
+	global $wpdb;
```

Comment:
I was looking at returning early only for new products when there isn't a category and tag set but by default all products get the `uncategorised` category 🤦 .

Also, I would like to bail early if there is no change in categories or tags when updating a product but couldn't find a way to identify changes.

Reply:
Those were my first thoughts when reviewing this PR, but I understand we don't have access to the previous categories/tags and because of that we have no way to know which products need to be updated, right?

But if that's the case, do we need to pass `$product_id` into this function? :thinking:

Reply:
Good catch! I had a call with Tom and decided to delete the transients only for the affected products ( old related products and new related products ) so function signature will change anyway. Thanks for the review!